### PR TITLE
Add R UDF Blog

### PR DESCRIPTION
**Pull request by SaltDonkey**
This PR adds a blog for R UDF.

### CODE REVIEW

File: content/en/blog/r-udf/index.md [hunk 1]
```diff
@@ -0,0 +1,192 @@
+---
+title: "Adding R UDF to Texera: The Journey"
+description: ""
+lead: "In this blog, we discuss the journey taken to implement support for R in Texera. Most notably, we allowed support for Table and Tuple API with R in Texera
+."
+date: 2024-07-06T08:00:00+00:00
```
Comment (Yicong-Huang) [inline content/en/blog/r-udf/index.md:6]: update the date to current date and time

File: content/en/blog/r-udf/index.md [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+lastmod: 2024-07-06T11:59:00+00:00
```
Comment (Yicong-Huang) [inline content/en/blog/r-udf/index.md:7]: update the date to current date and time

File: content/en/blog/r-udf/index.md [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+draft: false
+weight: 50
+contributors: ["Yicong Huang", "Kevin Wu" ]
+affiliations: ["Computer Science, UC Irvine"]
+---
+
+# Motivation
```
Comment (Yicong-Huang) [inline content/en/blog/r-udf/index.md:14]: use `##` for first level headers

File: content/en/blog/r-udf/index.md [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+
+The R programming language is widely used in many fields, such as bioinformatics, statistical analysis, and social science. Its popularity is driven by its capabilities and the growing demand for data-related work. Given its significance, our goal was to allow users to seamlessly integrate R into Texera workflows.
+
+We want to enhance Texera by extending its support to include R code and facilitating the creation of R user-defined functions (UDFs) for both the table and tuple APIs. Doing so would mirror the functionality of Python UDFs, allowing both experienced and new users of Texera to leverage R in their work.
+
+# Challenges
+The current support for R UDF in Texera utilizes Python User-Defined Functions and `rpy2` (a Python library to use R in Python) to incorporate R code. However, this approach has two limitations, compatability and performance.
+
+## Challenge 1: Compatibility
```
Comment (Yicong-Huang) [inline content/en/blog/r-udf/index.md:23]: use `###` for second level headers

File: content/en/blog/r-udf/index.md [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+
+# Our Proposed Design
+
+Our proposed architecture features a three-layered design, where the data flows from Java to Python, then to R, before returning back to Python and Java. An  illustration of this architecture is in *Figure 1* below.
+
+<figure style="margin: auto; text-align: center">
+<a>
+<img src="rudf_architecture.png"Fig 1" style="width:80%;">
+</a>
+<figcaption align=center style="text-align:center; font-style: italic">Figure 1: Figure 1: A concise overview of our proposed design for R UDF with all the libraries marked.</figcaption>
```
Comment (Yicong-Huang) [inline content/en/blog/r-udf/index.md:50]: `Figure 1` is duplicated twice

File: content/en/blog/r-udf/index.md [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+As a showcase of the successes of our proposed design, below in *Figure 2* is a sample workflow created by one of our users that exclusively uses R UDFs, all of which employed the tuple API.
+
+<figure style="margin: auto; text-align: center">
+<a>
+<img src="rudf_sample_workflow.png" alt="Fig 2" style="width:100%;">
+</a>
+<figcaption align=center style="text-align:center; font-style: italic">Figure 2: A sample workflow using R UDFs only. All of the operators are using the tuple API.</figcaption>
+</figure>
+
+This demo workflow was used in a research project intended to conduct Single Cell RNA analysis with R. The main driver for conducting the analysis came from the [`Seurat` ](https://satijalab.org/seurat/) package, which contains the eponymous `Seurat` object, an object that can store extensive data about single cells. In this workflow, the created `Seurat` object was frequently used as both input and output. Each step in the workflow utilized R UDFs to modify, extract, or inspect the `Seurat` object in some way, as can be seen from the labeled steps assigned to each R UDF operator.
```
Comment (Yicong-Huang) [inline content/en/blog/r-udf/index.md:83]: one extra space after `Seurat`

File: content/en/blog/r-udf/index.md [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+
+Instead, R Arrow Flight operates by invoking methods from the PyArrow Python package, including the code for running a Flight server in R. The most critical issue with this realization is because we wanted R UDF to be executed on the Flight server side. Due to the lack of a native R Flight server, this goal no longer was feasible.
+
+Regrettably, as of July 14, 2024, there was still no independent implementation of Arrow Flight. Currently, the R Arrow developers have no immediate plans to do so as well, as shown in their biweekly [meeting notes](https://docs.google.com/document/d/1nSIfJw8mfqtvScqvSVqmktpWff80pFmkqiZT7nTtiDo/edit#heading=h.k1ts4kvvl8jq). Consequently, we had to devise an alternative solution for R UDF that leverages Arrow and `rpy2` alone, eschewing Arrow Flight.
+
+# Our Proposed Design
+
+Our proposed architecture features a three-layered design, where the data flows from Java to Python, then to R, before returning back to Python and Java. An  illustration of this architecture is in *Figure 1* below.
+
+<figure style="margin: auto; text-align: center">
```
Comment (Yicong-Huang) [inline content/en/blog/r-udf/index.md:46]: for figures, try use similar format like the following example:

```
  <figure align="center" style="width: 100%; max-width: none;">
  <a href="https://raw.githubusercontent.com/Texera/texera.github.io/yiqundu/content/en/blog/flarum-discussion-integration/user-sych.gif">
    <img src="https://raw.githubusercontent.com/Texera/texera.github.io/yiqundu/content/en/blog/flarum-discussion-integration/user-sych.gif" alt="" style="width:100%">
  </a>
  <figcaption align = "center" style="font-size:1.2em;">
    <i>
      <b>Figure 3.</b> user logs into Texera.
    </i>
  </figcaption>
</figure>
```

File: content/en/blog/r-udf/index.md [hunk 8]
```diff
@@ -0,0 +1,192 @@
+---
+title: "Adding R UDF to Texera: The Journey"
+description: ""
+lead: "In this blog, we discuss the journey taken to implement support for R in Texera. Most notably, we allowed support for Table and Tuple API with R in Texera
+."
+date: 2024-07-06T08:00:00+00:00
+lastmod: 2024-07-06T11:59:00+00:00
+draft: false
+weight: 50
+contributors: ["Yicong Huang", "Kevin Wu" ]
```
Comment (Yicong-Huang) [inline content/en/blog/r-udf/index.md:10]: Put your name first!

**Review by Yicong-Huang — APPROVED** [review]
LGTM! Left some formatting suggestions.

File: content/en/blog/r-udf/index.md (hunk 8 above)
Comment (SaltDonkey) [reply content/en/blog/r-udf/index.md:10]: Fixed

File: content/en/blog/r-udf/index.md (hunk 3 above)
Comment (SaltDonkey) [reply content/en/blog/r-udf/index.md:14]: Fixed

File: content/en/blog/r-udf/index.md (hunk 4 above)
Comment (SaltDonkey) [reply content/en/blog/r-udf/index.md:23]: Fixed

File: content/en/blog/r-udf/index.md (hunk 6 above)
Comment (SaltDonkey) [reply content/en/blog/r-udf/index.md:83]: Fixed

File: content/en/blog/r-udf/index.md (hunk 5 above)
Comment (SaltDonkey) [reply content/en/blog/r-udf/index.md:50]: Fixed

File: content/en/blog/r-udf/index.md (hunk 2 above)
Comment (SaltDonkey) [reply content/en/blog/r-udf/index.md:7]: Fixed

File: content/en/blog/r-udf/index.md (hunk 1 above)
Comment (SaltDonkey) [reply content/en/blog/r-udf/index.md:6]: Fixed

**Pull request merged**